AC-454 adding visual clarity for selected menu options in the video player

clrux opened the pull request:
# [AC-454](https://openedx.atlassian.net/browse/AC-454)

This piece of work adds additional visual cues to selected menu options in the video player. Previously we were only using color to indicate selected options, but we need at least two identifying factors for good accessibility, especially in high contrast modes.

In addition to the color shift, I've also bolded the selected choice and placed a small border on the side of the option so it's exceedingly clear which option is chosen. The border is a redundant cue in case the bolder type or color shift isn't enough.

I also synced up behavior for the two menus (speed and language). Previously a user could change the transcript languages, but the menu would remain open, whereas choosing an option in the speed would update the selection and close the menu. Now both menus close on choice selection.

Tests already check for the presence of `is-active` on both of these menus.

## Visual updates

![screen shot 2016-05-10 at 2 57 31 pm](https://cloud.githubusercontent.com/assets/2112024/15158923/0bbdcd5a-16c0-11e6-9d74-f0c961b294ff.png)
_Normal display_

![img_20160511_082617](https://cloud.githubusercontent.com/assets/2112024/15180900/3ea1a13c-1752-11e6-902b-2ff2c7544968.jpg)
_High contrast display_

## Sandbox

https://clrux-ac-454.sandbox.edx.org/courses/course-v1:edX+JUS101+2016_T2/courseware/C_01/lecture_01/

## Reviewers

- [x] @cptvitamin
- [x] @muzaffaryousaf

clrux commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -210,21 +210,32 @@ function (Iterator) {
+                    .siblings("li[data-speed='" + speed + "']");
 
-                this.speedButton.find('.value').html(speed + 'x');
+                this.speedButton.find('.value').text(speed + 'x');
```
Safe templates.

clrux commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -704,7 +705,7 @@
                             'data-start': start[index],
                             'tabindex': 0
-                        }).html(text);
+                        }).text(text);
```
Safe templates.

clrux commented on common/lib/xmodule/xmodule/css/video/display.scss:
```diff
@@ -534,6 +534,8 @@
 
                 .speed-option,
                 .control-lang {
+                  @include border-left($baseline/10 solid rgb(14, 166, 236));
```
Same color as `color` below. Should probably be a variable?

muzaffaryousaf replied:
@clrux FYI, I'm just looking at the video code as i'm not expert in a11y or styling.

clrux replied:
Thanks @muzaffaryousaf that's fine. I'll have someone here review the styling. Focus on the video code itself.

clrux commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -210,21 +210,35 @@ function (Iterator) {
          */
         setSpeed: function (speed, silent, forceUpdate) {
-            if (speed !== this.currentSpeed || forceUpdate) {
+            if (speed !== this.state.speed || forceUpdate) {
```
`state.speed` is more reliable.

muzaffaryousaf replied:
Why it's more reliable when they are being set on same place.

clrux replied:
@muzaffaryousaf I removed the duplicate.

clrux commented:
jenkins run js

cahrens commented:
@clrux the Lahore TNL folks know the video code better, so I think they should be your second code reviewer. I have updated the checklist.

clrux commented:
@muzaffaryousaf Do you think you can get to this review sometime soon?

cptvitamin commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -210,20 +211,44 @@ function (Iterator) {
 
-                this.speedButton.find('.value').html(speed + 'x');
+                this.speedButton.find('.value').text(speed + 'x');
                 this.currentSpeed = speed;
```
Any reason to keep this around?

cptvitamin commented:
@clrux this functions as expected in regular and high contrast mode.  I think it makes a big improvement in UX, too.  Making this more accessible to low vision users may be aware of something we should be doing for non-visual users as well, and that is indicating what the current setting for these two controls is, and also indicating that (using screen reader text) in the menu itself.  Can you squeeze that into this PR?  

1. Looking at the src code I would change the approach for adding the screen reader help text to offscreen text instead of aria-label, that way you can just add a new span with the help text to the button: "Speed: 1.50x Press up to enter the speed..."
2. Add screen reader text to the currently selected menu item (class="is-active") that indicates it is the currently selected menu item: "1.50x selected"

muzaffaryousaf commented:
@clrux I'll take a look today.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -210,20 +211,44 @@ function (Iterator) {
+                this.speedButton.find('.value').text(speed + 'x');
                 this.currentSpeed = speed;
-
+                this.state.speed = speed;
```
Why we're setting `speed` to `this.state.speed` when it's already present on `this.currentSpeed` ?

clrux commented:
@cptvitamin Per your advice I've added some auditory context too. The part that covers removing the `aria-label` to `sr` text will be largely covered in [AC-406](https://github.com/edx/edx-platform/pull/12430) but I've gone ahead and done the following here:

* added a `title` to the base control that contains the current speed
* removed the `role="application"` as this was preventing the `title` from being read
* added `aria-pressed="true"` to the selected speed within in the expanded menu

clrux commented:
@cptvitamin Per your advice I've added some auditory context too. The part that covers removing the `aria-label` to `sr` text will be largely covered in [AC-406](https://github.com/edx/edx-platform/pull/12430) but I've gone ahead and done the following here:

* added a `title` to the base control that contains the current speed
* added `aria-pressed="true"` to the selected speed within in the expanded menu

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -109,6 +109,7 @@ function (Iterator) {
             speedsContainer.html(speedsList.join(''));
             this.speedLinks = new Iterator(speedsContainer.find('.speed-option'));
             this.state.el.find('.secondary-controls').prepend(this.el);
+            this.setActiveSpeed("1.0"); // default state on initial render
```
Why we need this ? It will highlight the wrong value when the speed is not "1.0". You can see in screenshot too.
![screen shot 2016-05-18 at 5 07 51 pm](https://cloud.githubusercontent.com/assets/2994509/15357913/1f23afd4-1d1b-11e6-8761-7da8470cb62d.png)

clrux replied:
@muzaffaryousaf This is set in the initial render of the page. This sets a default of 1.0x speed in case one isn't set. I tested this without issues; are you having trouble changing speeds?

muzaffaryousaf replied:
@clrux In attached screenshot the speed is `1.50x`  but it's highlighting `1.0`. Should't it highlight the given speed value ?

clrux replied:
Got it. I will take another look at this. Maybe I will check for the existence first and use that value if it's present, with a default backup to 1.0x. This may also be set as a cookie.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -213,17 +214,38 @@ function (Iterator) {
+        },
+        
+        resetActiveSpeed: function() {
+            var options = this.speedsContainer.find('li');
```
Can we call this `speedOptions`.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -213,17 +214,38 @@ function (Iterator) {
+        },
+        
+        setActiveSpeed: function(speed) {
+            var option = this.speedsContainer.find('li[data-speed="' + speed + '"]');
```
`speedOption`.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -241,10 +270,13 @@ function (Iterator) {
+                speed = $(el).data('speed');
+                
+            this.resetActiveSpeed();
+            this.setActiveSpeed(speed);
```
I did't get the purpose of these changes, can you shed some light on this.

clrux replied:
The work here is mainly front-end. In short, we needed to better indicate the selected state for high contrast screens. These new functions add necessary style classes to achieve this. More information is in the description at the top of this PR. Let me know if you have other questions.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/09_video_caption.js:
```diff
@@ -682,6 +682,7 @@
 
                         state.el.trigger('language_menu:change', [langCode]);
                         self.fetchCaption();
+                        self.closeLanguageMenu(e);
```
Was it not being closed earlier ?

clrux replied:
The language menu and speed menus behaved differently before. The speed menu would close when a selection was made but the language menu wouldn't. I wanted both menus to behave the same so I needed to make the language menu close when a choice was selected.

muzaffaryousaf replied:
Got it. Thanks for this.

muzaffaryousaf commented:
@clrux I left some comments, let me know when you're done addressing them. Thanks.

clrux commented:
Thanks @muzaffaryousaf. I believe I've addressed your feedback and questions. Please let me know if you have other questions. Be sure and check out the description of this PR at the top for more details too.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -109,6 +110,13 @@ function (Iterator) {
+                initialSpeed = this.state.speed;
+            } else {
+                // default if no speed is set
+                initialSpeed = "1.0";
```
I think it would be nice if we can use value from speeds instead of hardcoding this.

muzaffaryousaf replied:
Thoughts on this?

clrux replied:
If we were to use `speeds` we'd still want to find "1.0" in that object, so in a sense it would still be hard-coded. I'd like the default to be "normal" speed (1.0x). What would you suggest here?

clrux replied:
Yep. It looks like `render()` receives `speed` as a variable, and it's always set, so I'm using it instead of re-setting variables.

muzaffaryousaf commented:
@clrux Did you updated the sandbox after making the changes, i'm still seeing the old behavior for highlighting the speed according to it's value.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -93,13 +93,13 @@ function (Iterator) {
          * @param {array} speeds List of speeds available for the player.
          */
-        render: function (speeds) {
+        render: function (speeds, speed) {
```
We need to add doc string for `speed` parameter just like `speeds`.

muzaffaryousaf replied:
Can we add js test to check if the expected speed option get highlighted and being set.

muzaffaryousaf replied:
Can we also name this `currentSpeed` just to avoid confusion with temp variable that we're using in map method.

clrux replied:
@muzaffaryousaf I've added a doc string and updated the variable name. We already have tests that check for the functionality. The test checks for the appropriate class, but it doesn't check the styles. I think this is OK, as long as the correct markup is present.

I'm updating the sandbox right now.

muzaffaryousaf commented:
@clrux I'm done with another pass and left some comments. Let me know when you're done with addressing them.

clrux commented:
jenkins run bokchoy

clrux commented:
@muzaffaryousaf I've addressed your new comments and updated the sandbox.

muzaffaryousaf commented:
@clrux LGTM 👍 once we get green build. Thanks.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/video/08_video_speed_control.js:
```diff
@@ -213,17 +215,38 @@ function (Iterator) {
 
                 if (!silent) {
-                    this.el.trigger('speedchange', [speed, this.state.speed]);
+                    this.el.trigger('speedchange', [speed, this.currentSpeed]);
```
One more thing we don't need this change here. Do we?

muzaffaryousaf replied:
Let's not change that's not necessary. Can we use `this.state.speed` again. Thanks.

clrux commented:
jenkins run bokchoy

clrux commented:
@cptvitamin Do you think you could review this one sometime next week?

clrux commented:
jenkins run bokchoy

cptvitamin commented:
@clrux this looks good, but can you apply the same changes to the Language button as well?  Thanks

clrux commented:
jenkins run bokchoy

cptvitamin commented:
fix the failing test and merge this 👍

clrux commented:
jenkins run bokchoy
